src/Controller/Component/ModulesComponent.php

```diff
@@ -145,10 +145,39 @@ function (&$data, $key) use ($metaModules) {
             array_diff_key($metaModules, $modules),
             $pluginModules
         );
+        $this->modules = $this->modulesByRoleConfig($this->modules);
 
         return $this->modules;
     }
 
+    /**
+     * This filters modules and apply 'Role' config by user role, if any.
+     * Module can be "hidden": remove from $modules.
+     * Module can be "readonly": add "'readonly' => true" to module.
+     *
+     * @param array $modules The modules
+     * @return array
+     */
+    protected function modulesByRoleConfig(array $modules): array
+    {
+        $user = $this->getController()->Auth->user();
+        $role = (string)Hash::get($user, 'roles.0');
```

Comment:
We should check against all roles not just the first one. 
We could limit to a list of roles via an `array_intersect()` of user roles and roles defined in configuration.

Reply:
`AccessControl` is already a configuration.
I would avoid too much configuration... `AccessControl` configuration keys could do the task: providing user roles